sdk/storage/azure_storage_blob/src/streams/partitioned_stream.rs

```diff
@@ -0,0 +1,199 @@
+use std::{
+    mem,
+    num::NonZero,
+    pin::{pin, Pin},
+    task::Poll,
+};
+
+use azure_core::stream::SeekableStream;
+use bytes::Bytes;
+use futures::{ready, stream::FusedStream, AsyncRead, Stream};
+
+type AzureResult<T> = azure_core::Result<T>;
+
+pub(crate) struct PartitionedStream {
+    inner: Box<dyn SeekableStream>,
+    buf: Vec<u8>,
+    partition_len: usize,
+    buf_offset: usize,
+    total_read: usize,
+    inner_complete: bool,
+}
+
+impl PartitionedStream {
+    pub(crate) fn new(inner: Box<dyn SeekableStream>, partition_len: NonZero<usize>) -> Self {
+        let partition_len = partition_len.get();
+        Self {
+            buf: vec![0u8; std::cmp::min(partition_len, inner.len())],
+            inner,
+            partition_len,
+            buf_offset: 0,
+            total_read: 0,
+            inner_complete: false,
+        }
+    }
+
+    fn take(&mut self) -> Vec<u8> {
+        let mut ret = mem::replace(
+            &mut self.buf,
+            vec![0u8; std::cmp::min(self.partition_len, self.inner.len() - self.total_read)],
+        );
+        ret.truncate(self.buf_offset);
+        self.buf_offset = 0;
+        ret
+    }
```

Comment:
You're already taking a dependency on `Bytes`. Use it. It already has all the functionality for this. At the very least, don't repeat the calculation of `buf.len()` that you'd already have computed during construction. `Bytes` has already been well-tested. I see potential faults here.